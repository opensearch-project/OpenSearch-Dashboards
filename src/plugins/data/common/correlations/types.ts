/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { SavedObject, SavedObjectAttributes } from '../../../../core/types';

// @experimental This schema is experimental and might change in future releases.
export interface CorrelationSavedObjectAttributes extends SavedObjectAttributes {
  correlationType: string;

  version: string;

  /*
   * Using any[] for entities to allow flexible correlation structures.
   * This enables different correlation types and use cases to define their own entity schemas without strict typing constraints.
   * Each entity can contain different fields like correlatedFields, meta, datasource info.
   * @experimental - The any[] type is experimental and may be replaced with more specific types in future releases.
   */
  entities: any[];
}

export interface CorrelationSavedObject extends SavedObject<CorrelationSavedObjectAttributes> {
  type: 'correlations';
}
