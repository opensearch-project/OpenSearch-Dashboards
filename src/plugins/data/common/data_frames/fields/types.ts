/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export interface IFieldType {
  name: string;
  type: string;
  values: any[];
  count?: number;
  aggregatable?: boolean;
  filterable?: boolean;
  searchable?: boolean;
  sortable?: boolean;
  visualizable?: boolean;
  displayName?: string;
  format?: any;
}
