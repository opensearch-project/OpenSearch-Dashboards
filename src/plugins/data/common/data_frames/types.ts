/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { SearchResponse } from 'elasticsearch';
import { IFieldType } from './fields';

export * from './_df_cache';

/** @public **/
export enum DATA_FRAME_TYPES {
  DEFAULT = 'data_frame',
  POLLING = 'data_frame_polling',
}

export interface DataFrameService {
  get: () => IDataFrame | undefined;
  set: (dataFrame: IDataFrame) => Promise<void>;
  clear: () => void;
}

/**
 * A data frame is a two-dimensional labeled data structure with columns of potentially different types.
 */
export interface IDataFrame {
  type?: DATA_FRAME_TYPES.DEFAULT;
  name?: string;
  schema?: Array<Partial<IFieldType>>;
  meta?: Record<string, any>;
  fields: IFieldType[];
  size: number;
}

/**
 * An aggregation is a process where the values of multiple rows are grouped together to form a single summary value.
 */
export interface DataFrameAgg {
  value: number;
}

/**
 * A bucket aggregation is a type of aggregation that creates buckets or sets of data.
 */
export interface DataFrameBucketAgg extends DataFrameAgg {
  key: string;
}

/**
 * This configuration is used to define how the aggregation should be performed.
 */
export interface DataFrameAggConfig {
  id: string;
  type: string;
  field?: string;
  order?: Record<string, string>;
  size?: number;
  date_histogram?: {
    field: string;
    fixed_interval?: string;
    calendar_interval?: string;
    time_zone: string;
    min_doc_count: number;
  };
  avg?: {
    field: string;
  };
  cardinality?: {
    field: string;
  };
  terms?: {
    field: string;
    size: number;
    order: Record<string, string>;
  };
  aggs?: Record<string, DataFrameAggConfig>;
}

export interface PartialDataFrame extends Omit<IDataFrame, 'fields' | 'size'> {
  fields: Array<Partial<IFieldType>>;
}

/**
 * To be utilize with aggregations and will map to buckets
 * Plugins can get the aggregated value by their own logic
 * Setting to null will disable the aggregation if plugin wishes
 * In future, if the plugin doesn't intentionally set the value to null,
 * we can calculate the value based on the fields.
 */
// TODO: handle composite
export interface IDataFrameWithAggs extends IDataFrame {
  aggs: Record<string, DataFrameAgg | DataFrameBucketAgg | DataFrameBucketAgg[]>;
}

export interface IDataFrameResponse extends SearchResponse<any> {
  type: DATA_FRAME_TYPES;
  body: IDataFrame | IDataFrameWithAggs | IDataFrameError;
  took: number;
}

export interface IDataFrameError {
  error: Error;
}
