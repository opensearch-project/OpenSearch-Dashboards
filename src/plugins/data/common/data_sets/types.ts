/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

/** @public **/
export enum SIMPLE_DATA_SOURCE_TYPES {
  DEFAULT = 'data-source',
  EXTERNAL = 'external-source',
}

/** @public **/
export enum SIMPLE_DATA_SET_TYPES {
  INDEX_PATTERN = 'index-pattern',
  TEMPORARY = 'temporary',
  TEMPORARY_ASYNC = 'temporary-async',
}

export interface SimpleObject {
  id: string;
  title?: string;
  dataSourceRef?: SimpleDataSource;
}

export interface SimpleDataSource {
  id: string;
  name: string;
  indices?: SimpleObject[];
  tables?: SimpleObject[];
  type: SIMPLE_DATA_SOURCE_TYPES;
}

export interface SimpleDataSet extends SimpleObject {
  fields?: any[];
  timeFieldName?: string;
  timeFields?: any[];
  type?: SIMPLE_DATA_SET_TYPES;
}
