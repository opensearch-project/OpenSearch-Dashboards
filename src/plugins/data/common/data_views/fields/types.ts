/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { DataViewFieldSpec, IDataViewFieldSubType, DataView } from '../../data_views';

export interface IDataViewFieldType {
  name: string;
  type: string;
  script?: string;
  lang?: string;
  count?: number;
  // esTypes might be undefined on old index patterns that have not been refreshed since we added
  // this prop. It is also undefined on scripted fields.
  esTypes?: string[];
  aggregatable?: boolean;
  filterable?: boolean;
  searchable?: boolean;
  sortable?: boolean;
  visualizable?: boolean;
  readFromDocValues?: boolean;
  scripted?: boolean;
  subType?: IDataViewFieldSubType;
  displayName?: string;
  format?: any;
  toSpec?: (options?: {
    getFormatterForField?: DataView['getFormatterForField'];
  }) => DataViewFieldSpec;
}
