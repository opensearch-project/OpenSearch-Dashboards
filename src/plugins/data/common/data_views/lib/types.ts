/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const ILLEGAL_CHARACTERS_KEY = 'ILLEGAL_CHARACTERS';
export const CONTAINS_SPACES_KEY = 'CONTAINS_SPACES';
export const ILLEGAL_CHARACTERS_VISIBLE = ['\\', '/', '?', '"', '<', '>', '|'];
export const ILLEGAL_CHARACTERS = ILLEGAL_CHARACTERS_VISIBLE.concat(' ');
