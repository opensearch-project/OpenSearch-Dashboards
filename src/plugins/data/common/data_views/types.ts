/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  IIndexPattern,
  IndexPatternAttributes,
  OnNotification,
  OnError,
  OnUnsupportedTimePattern,
  UiSettingsCommon,
  SavedObjectsClientCommonFindArgs,
  SavedObjectsClientCommon,
  GetFieldsOptions,
  IIndexPatternsApiClient,
  SavedObject,
  AggregationRestrictions,
  IFieldSubType,
  TypeMeta,
  FieldSpecConflictDescriptions,
  FieldSpecExportFmt,
  FieldSpec,
  IndexPatternFieldMap,
  SavedObjectReference,
  IndexPatternSpec,
  SourceFilter,
} from '../index_patterns/types';
import { Dataset } from '../datasets/types';

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export interface IDataView extends IIndexPattern {
  displayName?: string;
  description?: string;
  type?: string;
  dataSourceRef?: DataViewSavedObjectReference;

  /**
   * @experimental This method is experimental and may change in future versions
   */
  initializeDataSourceRef?(): Promise<void>;

  /**
   * Converts a DataView to a serializable Dataset object
   * Maps dataSourceRef and includes only essential properties for backward compatibility
   * @experimental This method is experimental and may change in future versions
   */
  toDataset?(): Promise<Dataset>;
}

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewAttributes = IndexPatternAttributes;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewOnNotification = OnNotification;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewOnError = OnError;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewOnUnsupportedTimePattern = OnUnsupportedTimePattern;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewUiSettingsCommon = UiSettingsCommon;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewSavedObjectsClientCommonFindArgs = SavedObjectsClientCommonFindArgs;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewSavedObjectsClientCommon = SavedObjectsClientCommon;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewGetFieldsOptions = GetFieldsOptions;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type IDataViewsApiClient = IIndexPatternsApiClient;

export type { SavedObject };

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewAggregationRestrictions = AggregationRestrictions;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type IDataViewFieldSubType = IFieldSubType;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewTypeMeta = TypeMeta;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewFieldSpecConflictDescriptions = FieldSpecConflictDescriptions;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewFieldSpecExportFmt = FieldSpecExportFmt;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewFieldSpec = FieldSpec;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewFieldMap = IndexPatternFieldMap;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewSavedObjectReference = SavedObjectReference;

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewSpec = IndexPatternSpec & {
  dataSourceRef?: DataViewSavedObjectReference;
};

/**
 * @experimental DataView functionality is experimental and may change in future versions
 */
export type DataViewSourceFilter = SourceFilter;
