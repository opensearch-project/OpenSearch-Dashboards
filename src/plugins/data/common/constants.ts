/*
 * SPDX-License-Identifier: Apache-2.0
 *
 * The OpenSearch Contributors require contributions made to
 * this file be licensed under the Apache-2.0 license or a
 * compatible open source license.
 *
 * Any modifications Copyright OpenSearch Contributors. See
 * GitHub history for details.
 */

import { DATA_STRUCTURE_META_TYPES, DataStructure } from './types';

/*
 * Licensed to Elasticsearch B.V. under one or more contributor
 * license agreements. See the NOTICE file distributed with
 * this work for additional information regarding copyright
 * ownership. Elasticsearch B.V. licenses this file to you under
 * the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

export const DEFAULT_DATA = {
  STRUCTURES: {
    ROOT: {
      id: 'ROOT',
      title: 'Data',
      type: 'ROOT',
      meta: {
        type: DATA_STRUCTURE_META_TYPES.FEATURE,
        icon: { type: 'folderOpen' },
        tooltip: 'Root Data Structure',
      },
    } as DataStructure,
    LOCAL_DATASOURCE: {
      id: '',
      title: 'Default Cluster',
      type: 'DATA_SOURCE',
    },
  },

  SET_TYPES: {
    INDEX_PATTERN: 'INDEX_PATTERN',
    INDEX: 'INDEXES',
  },

  SOURCE_TYPES: {
    OPENSEARCH: 'OpenSearch',
    LEGACY: 'LEGACY',
  },
};

export const DEFAULT_QUERY_LANGUAGE = 'kuery';

export const DEFAULT_QUERY = {
  LANGUAGE: DEFAULT_QUERY_LANGUAGE,
  DATASET: {
    TYPE: DEFAULT_DATA.SET_TYPES.INDEX_PATTERN,
    DATASOURCE: {
      TYPE: DEFAULT_DATA.SOURCE_TYPES.OPENSEARCH,
    },
  },
};

export const UI_SETTINGS = {
  META_FIELDS: 'metaFields',
  DOC_HIGHLIGHT: 'doc_table:highlight',
  QUERY_STRING_OPTIONS: 'query:queryString:options',
  QUERY_ALLOW_LEADING_WILDCARDS: 'query:allowLeadingWildcards',
  SEARCH_QUERY_LANGUAGE: 'search:queryLanguage',
  SORT_OPTIONS: 'sort:options',
  COURIER_IGNORE_FILTER_IF_FIELD_NOT_IN_INDEX: 'courier:ignoreFilterIfFieldNotInIndex',
  COURIER_SET_REQUEST_PREFERENCE: 'courier:setRequestPreference',
  COURIER_CUSTOM_REQUEST_PREFERENCE: 'courier:customRequestPreference',
  COURIER_MAX_CONCURRENT_SHARD_REQUESTS: 'courier:maxConcurrentShardRequests',
  COURIER_BATCH_SEARCHES: 'courier:batchSearches',
  SEARCH_INCLUDE_FROZEN: 'search:includeFrozen',
  SEARCH_TIMEOUT: 'search:timeout',
  SEARCH_INCLUDE_ALL_FIELDS: 'search:includeAllFields',
  SEARCH_MAX_RECENT_DATASETS: 'search:maxRecentDatasets',
  HISTOGRAM_BAR_TARGET: 'histogram:barTarget',
  HISTOGRAM_MAX_BARS: 'histogram:maxBars',
  HISTORY_LIMIT: 'history:limit',
  SHORT_DOTS_ENABLE: 'shortDots:enable',
  FORMAT_DEFAULT_TYPE_MAP: 'format:defaultTypeMap',
  FORMAT_NUMBER_DEFAULT_PATTERN: 'format:number:defaultPattern',
  FORMAT_PERCENT_DEFAULT_PATTERN: 'format:percent:defaultPattern',
  FORMAT_BYTES_DEFAULT_PATTERN: 'format:bytes:defaultPattern',
  FORMAT_CURRENCY_DEFAULT_PATTERN: 'format:currency:defaultPattern',
  FORMAT_NUMBER_DEFAULT_LOCALE: 'format:number:defaultLocale',
  TIMEPICKER_REFRESH_INTERVAL_DEFAULTS: 'timepicker:refreshIntervalDefaults',
  TIMEPICKER_QUICK_RANGES: 'timepicker:quickRanges',
  TIMEPICKER_TIME_DEFAULTS: 'timepicker:timeDefaults',
  INDEXPATTERN_PLACEHOLDER: 'indexPattern:placeholder',
  FILTERS_PINNED_BY_DEFAULT: 'filters:pinnedByDefault',
  FILTERS_EDITOR_SUGGEST_VALUES: 'filterEditor:suggestValues',
  QUERY_ENHANCEMENTS_ENABLED: 'query:enhancements:enabled',
  QUERY_DATAFRAME_HYDRATION_STRATEGY: 'query:dataframe:hydrationStrategy',
  SEARCH_QUERY_LANGUAGE_BLOCKLIST: 'search:queryLanguageBlocklist',
  NEW_HOME_PAGE: 'home:useNewHomePage',
} as const;
