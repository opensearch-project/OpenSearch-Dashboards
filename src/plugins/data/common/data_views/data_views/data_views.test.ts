/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { defaults } from 'lodash';
import { DataViewsService, DataView } from '.';
import { fieldFormatsMock } from '../../field_formats/mocks';
import { stubbedSavedObjectIndexPattern } from '../../../../../fixtures/stubbed_saved_object_index_pattern';
import { DataViewUiSettingsCommon, DataViewSavedObjectsClientCommon, SavedObject } from '../types';
import { UI_SETTINGS } from '../../constants';

const createFieldsFetcher = jest.fn().mockImplementation(() => ({
  getFieldsForWildcard: jest.fn().mockImplementation(() => {
    return new Promise((resolve) => resolve([]));
  }),
  every: jest.fn(),
}));

const fieldFormats = fieldFormatsMock;
let object: any = {};

function setDocsourcePayload(id: string | null, providedPayload: any) {
  object = defaults(providedPayload || {}, stubbedSavedObjectIndexPattern(id));
}

describe('DataViews', () => {
  let indexPatterns: DataViewsService;
  let savedObjectsClient: DataViewSavedObjectsClientCommon;
  const uiSettingsGet = jest.fn();

  beforeEach(() => {
    const indexPatternObj = { id: 'id', version: 'a', attributes: { title: 'title' } };
    savedObjectsClient = {} as DataViewSavedObjectsClientCommon;
    savedObjectsClient.find = jest.fn(
      () => Promise.resolve([indexPatternObj]) as Promise<Array<SavedObject<any>>>
    );
    savedObjectsClient.delete = jest.fn(() => Promise.resolve({}) as Promise<any>);
    savedObjectsClient.create = jest.fn();
    savedObjectsClient.get = jest.fn().mockImplementation(async (type, id) => ({
      id: object.id,
      version: object.version,
      attributes: object.attributes,
    }));
    savedObjectsClient.update = jest
      .fn()
      .mockImplementation(async (type, id, body, { version }) => {
        if (object.version !== version) {
          throw new Object({
            res: {
              status: 409,
            },
          });
        }
        object.attributes.title = body.title;
        object.version += 'a';
        return {
          id: object.id,
          version: object.version,
        };
      });

    uiSettingsGet.mockClear();
    uiSettingsGet.mockReturnValue(Promise.resolve(false));

    indexPatterns = new DataViewsService({
      uiSettings: ({
        get: uiSettingsGet,
        getAll: () => {},
      } as any) as DataViewUiSettingsCommon,
      savedObjectsClient: (savedObjectsClient as unknown) as DataViewSavedObjectsClientCommon,
      apiClient: createFieldsFetcher(),
      fieldFormats,
      onNotification: () => {},
      onError: () => {},
      onRedirectNoDataView: () => {},
      onUnsupportedTimePattern: () => {},
    });
  });

  test('does cache gets for the same id', async () => {
    const id = '1';
    setDocsourcePayload(id, {
      id: 'foo',
      version: 'foo',
      attributes: {
        title: 'something',
      },
    });

    const indexPattern = await indexPatterns.get(id);

    expect(indexPattern).toBeDefined();
    expect(indexPattern).toBe(await indexPatterns.get(id));
  });

  test('savedObjectCache pre-fetches only title', async () => {
    expect(await indexPatterns.getIds()).toEqual(['id']);
    expect(savedObjectsClient.find).toHaveBeenCalledWith({
      type: 'index-pattern',
      fields: ['title'],
      perPage: 10000,
    });
  });

  test('caches saved objects', async () => {
    await indexPatterns.getIds();
    await indexPatterns.getTitles();
    expect(savedObjectsClient.find).toHaveBeenCalledTimes(1);
  });

  test('can refresh the saved objects caches', async () => {
    await indexPatterns.getIds();
    await indexPatterns.getTitles(true);
    expect(savedObjectsClient.find).toHaveBeenCalledTimes(2);
  });

  test('deletes the index pattern', async () => {
    const id = '1';
    const indexPattern = await indexPatterns.get(id);

    expect(indexPattern).toBeDefined();
    await indexPatterns.delete(id);
    expect(indexPattern).not.toBe(await indexPatterns.get(id));
  });

  test('should handle version conflicts', async () => {
    setDocsourcePayload(null, {
      id: 'foo',
      version: 'foo',
      attributes: {
        title: 'something',
      },
    });

    // Create a normal index patterns
    const pattern = await indexPatterns.get('foo');

    expect(pattern.version).toBe('fooa');
    indexPatterns.clearCache();

    // Create the same one - we're going to handle concurrency
    const samePattern = await indexPatterns.get('foo');

    expect(samePattern.version).toBe('fooaa');

    // This will conflict because samePattern did a save (from refreshFields)
    // but the resave should work fine
    pattern.title = 'foo2';
    await indexPatterns.updateSavedObject(pattern);

    // This should not be able to recover
    samePattern.title = 'foo3';

    let result;
    try {
      await indexPatterns.updateSavedObject(samePattern);
    } catch (err) {
      result = err;
    }

    expect(result.res.status).toBe(409);
  });

  test('create', async () => {
    const title = 'opensearch-dashboards-*';
    indexPatterns.refreshFields = jest.fn();

    const indexPattern = await indexPatterns.create({ title }, true);
    expect(indexPattern).toBeInstanceOf(DataView);
    expect(indexPattern.title).toBe(title);
    expect(indexPatterns.refreshFields).not.toBeCalled();

    await indexPatterns.create({ title });
    expect(indexPatterns.refreshFields).toBeCalled();
  });

  test('createAndSave', async () => {
    const title = 'opensearch-dashboards-*';
    indexPatterns.createSavedObject = jest.fn();
    indexPatterns.setDefault = jest.fn();
    await indexPatterns.createAndSave({ title });
    expect(indexPatterns.createSavedObject).toBeCalled();
    expect(indexPatterns.setDefault).toBeCalled();
  });

  test('savedObjectToSpec', () => {
    const savedObject = {
      id: 'id',
      version: 'version',
      attributes: {
        title: 'opensearch-dashboards-*',
        timeFieldName: '@timestamp',
        fields: '[]',
        sourceFilters: '[{"value":"item1"},{"value":"item2"}]',
        fieldFormatMap: '{"field":{}}',
        typeMeta: '{}',
        type: '',
      },
      type: 'index-pattern',
      references: [],
    };

    expect(indexPatterns.savedObjectToSpec(savedObject)).toMatchSnapshot();
  });

  test('savedObjectToSpecWithDataSource', () => {
    const savedObject = {
      id: 'id',
      version: 'version',
      attributes: {
        title: 'opensearch-dashboards-*',
        timeFieldName: '@timestamp',
        fields: '[]',
        sourceFilters: '[{"value":"item1"},{"value":"item2"}]',
        fieldFormatMap: '{"field":{}}',
        typeMeta: '{}',
        type: '',
      },
      type: 'index-pattern',
      references: [
        {
          id: 'id',
          type: 'data-source',
          name: 'dataSource',
        },
      ],
    };

    expect(indexPatterns.savedObjectToSpec(savedObject)).toMatchSnapshot();
  });

  test('correctly detects long-numerals support', async () => {
    expect(await indexPatterns.isLongNumeralsSupported()).toBe(false);

    uiSettingsGet.mockImplementation((key: string) =>
      Promise.resolve(key === UI_SETTINGS.DATA_WITH_LONG_NUMERALS ? true : undefined)
    );
    expect(await indexPatterns.isLongNumeralsSupported()).toBe(true);
  });
});
