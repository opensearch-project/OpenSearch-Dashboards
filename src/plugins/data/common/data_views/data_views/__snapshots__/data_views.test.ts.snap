// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataViews savedObjectToSpec 1`] = `
Object {
  "dataSourceRef": undefined,
  "description": undefined,
  "displayName": undefined,
  "fields": Object {},
  "id": "id",
  "intervalName": undefined,
  "sourceFilters": Array [
    Object {
      "value": "item1",
    },
    Object {
      "value": "item2",
    },
  ],
  "timeFieldName": "@timestamp",
  "title": "opensearch-dashboards-*",
  "type": "",
  "typeMeta": Object {},
  "version": "version",
}
`;

exports[`DataViews savedObjectToSpecWithDataSource 1`] = `
Object {
  "dataSourceRef": Object {
    "id": "id",
    "name": "dataSource",
    "type": "data-source",
  },
  "description": undefined,
  "displayName": undefined,
  "fields": Object {},
  "id": "id",
  "intervalName": undefined,
  "sourceFilters": Array [
    Object {
      "value": "item1",
    },
    Object {
      "value": "item2",
    },
  ],
  "timeFieldName": "@timestamp",
  "title": "opensearch-dashboards-*",
  "type": "",
  "typeMeta": Object {},
  "version": "version",
}
`;
